src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,293 @@
+    /// </summary>
+    private readonly Channel<TopicMessage> topicMessagesChannel = Channel.CreateUnbounded<TopicMessage>();
+    /// <summary>
+    /// Maintains the various acknowledgements for each message.
+    /// </summary>
+    private readonly Channel<TopicAcknowledgement> acknowledgementsChannel = Channel.CreateUnbounded<TopicAcknowledgement>();
+    /// <summary>
+    /// The stream connection between this instance and the Dapr sidecar.
+    /// </summary>
+    private AsyncDuplexStreamingCall<P.SubscribeTopicEventsRequestAlpha1, P.SubscribeTopicEventsResponseAlpha1>? clientStream;
+    /// <summary>
+    /// Used to ensure thread-safe operations against the stream.
+    /// </summary>
+    private readonly SemaphoreSlim semaphore = new(1, 1);
+    /// <summary>
+    /// The handler delegate responsible for processing the topic messages.
+    /// </summary>
+    private readonly TopicMessageHandler messageHandler;
+    /// <summary>
+    /// A reference to the DaprClient instance.
+    /// </summary>
+    private readonly P.Dapr.DaprClient client;
+    /// <summary>
+    /// Flag that prevents the developer from accidentally initializing the subscription more than once from the same receiver.
+    /// </summary>
+    private bool hasInitialized;
+    /// <summary>
+    /// Flag that ensures the instance is only disposed a single time.
+    /// </summary>
+    private bool isDisposed;
+
+    /// <summary>
+    /// Constructs a new instance of a <see cref="PublishSubscribeReceiver"/> instance.
+    /// </summary>
+    /// <param name="pubSubName">The name of the Dapr Publish/Subscribe component.</param>
+    /// <param name="topicName">The name of the topic to subscribe to.</param>
+    /// <param name="options">Options allowing the behavior of the receiver to be configured.</param>
+    /// <param name="handler">The delegate reflecting the action to take upon messages received by the subscription.</param>
+    /// <param name="client">A reference to the DaprClient instance.</param>
+    internal PublishSubscribeReceiver(string pubSubName, string topicName, DaprSubscriptionOptions options, TopicMessageHandler handler, P.Dapr.DaprClient client)
+    {
+        this.client = client;
+        this.pubSubName = pubSubName;
+        this.topicName = topicName;
+        this.options = options;
+        this.messageHandler = handler;
+    }
+
+    /// <summary>
+    /// Dynamically subscribes to messages on a PubSub topic provided by the Dapr sidecar.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing messages provided by the sidecar.</returns>
+    public async Task SubscribeAsync(CancellationToken cancellationToken = default)
+    {
+        //Prevents the receiver from performing the subscribe operation more than once (as the multiple initialization messages would cancel the stream).
```

Comment:
If the intent is that `SubscribeAsync()` is called once and only once, and there's nothing of any consequence that  the user can do between `Register()` and `SubscribeAsync()` (e.g. setup listeners, as that's done in the `Register()` call itself), then perhaps `SubscribeAsync()` should just be a private, implicit call by `Register[Async]()`?  In that case, `Register()` can just become `SubscribeAsync()` which returns an `IAsyncDisposable`.  Messages automatically start flowing when the method completes and stops when returned the disposable is called.

Reply:
Done + updated example to reflect the change